crates/driver/src/domain/competition/mod.rs

```diff
@@ -172,13 +172,48 @@ impl Competition {
         }
 
         // Pick the best-scoring settlement.
-        let (score, settlement) = scores
+        let (mut score, settlement) = scores
             .into_iter()
             .max_by_key(|(score, _)| score.to_owned())
             .map(|(score, settlement)| (Solved { score }, settlement))
             .unzip();
 
-        *self.settlement.lock().unwrap() = settlement;
+        *self.settlement.lock().unwrap() = settlement.clone();
+
+        let settlement = match settlement {
+            Some(settlement) => settlement,
+            // Don't wait for the deadline because we can't produce a solution anyway.
+            None => return Ok(score),
+        };
+
+        // Re-simulate the solution on every new block until the deadline ends to make
+        // sure we actually submit a working solution close to when the winner
+        // gets picked by the procotol.
+        if let Ok(deadline) = auction.deadline().timeout() {
+            let score_ref = &mut score;
+            let wait_and_resimulate = async move {
+                let mut stream =
+                    ethrpc::current_block::into_stream(self.eth.current_block().clone());
+                while let Some(block) = stream.next().await {
+                    match settlement.score(&self.eth, auction, &self.mempools.revert_protection()) {
```

Comment:
Or the simulation and scoring should be extracted from the encoding, so we can call them again on an already encoded settlement

Reply:
Good catch. 👍 
This code is getting a bit convoluted. It seems like the abstractions don't fit very well for the new use case.
I decided to only check whether the solution still works instead of re-encoding and re-calculating the fee.
As I don't expect either of them to make a huge difference. In the long term it might be nice to get this logic perfect but I would not want it to delay testing on prod unnecessarily.